### PR TITLE
[SPARK-12042] Python API for mllib.stat.test.StreamingTest

### PR DESCRIPTION
**Pull request by yinxusen**
## What changes were proposed in this pull request?

The patch adds python API for mllib.stat.test.StreamingTest under JIRA https://issues.apache.org/jira/browse/SPARK-12042.

Note that for `StreamingTestResult`, unlike other test results in Python, I define it as a normal Python class which doesn't extend from `TestResult` with a `_java_obj` in it.
## How was this patch tested?

The patch is tested with Python unit test.

### CODE REVIEW

**Comment by mengxr** [comment]
cc: @feynmanliang

**Comment by yinxusen** [comment]
test it please

**Comment by yinxusen** [comment]
cc @feynmanliang

**Comment by zsxwing** [comment]
Any updates to this PR?

**Comment by yinxusen** [comment]
Ping @mengxr @feynmanliang @yanboliang

**Comment by feynmanliang** [comment]
I'll review this tonight

**Comment by feynmanliang** [comment]
Apologies for the delay, I am traveling but I'll get this done this weekend.

File: examples/src/main/python/mllib/streaming_test_example.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""
+Create a DStream that contains several RDDs to show the StreamingTest of PySpark.
+"""
```
Comment (feynmanliang) [inline examples/src/main/python/mllib/streaming_test_example.py]: Seems like other examples are including a `from __future__ import print_function` here

File: examples/src/main/python/mllib/streaming_test_example.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+from pyspark import SparkContext
+from pyspark.streaming import StreamingContext
+from pyspark.mllib.stat.test import BinarySample, StreamingTest
+
+if __name__ == "__main__":
+
+    sc = SparkContext(appName="PythonStreamingTestExample")
+    ssc = StreamingContext(sc, 1)
+
+    checkpoint_path = tempfile.mkdtemp()
```
Comment (feynmanliang) [inline examples/src/main/python/mllib/streaming_test_example.py]: Is this necessary?

File: examples/src/main/python/mllib/streaming_test_example.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import time
+import tempfile
+from shutil import rmtree
+
+from pyspark import SparkContext
+from pyspark.streaming import StreamingContext
+from pyspark.mllib.stat.test import BinarySample, StreamingTest
+
+if __name__ == "__main__":
+
```
Comment (feynmanliang) [inline examples/src/main/python/mllib/streaming_test_example.py]: nit: don't include newline here

File: examples/src/main/python/mllib/streaming_test_example.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+"""
+Create a DStream that contains several RDDs to show the StreamingTest of PySpark.
+"""
+import time
+import tempfile
+from shutil import rmtree
+
+from pyspark import SparkContext
+from pyspark.streaming import StreamingContext
+from pyspark.mllib.stat.test import BinarySample, StreamingTest
```
Comment (feynmanliang) [inline examples/src/main/python/mllib/streaming_test_example.py]: `$example on$` and `$example off` appear to be used in other examples, though I'm not sure why myself

File: examples/src/main/python/mllib/streaming_test_example.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+if __name__ == "__main__":
+
+    sc = SparkContext(appName="PythonStreamingTestExample")
+    ssc = StreamingContext(sc, 1)
+
+    checkpoint_path = tempfile.mkdtemp()
+    ssc.checkpoint(checkpoint_path)
+
+    # Create the queue through which RDDs can be pushed to a QueueInputDStream.
+    rdd_queue = []
```
Comment (feynmanliang) [inline examples/src/main/python/mllib/streaming_test_example.py]: use `camelCase` to be consistent with other examples

File: examples/src/main/python/mllib/streaming_test_example.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    # Create the QueueInputDStream and use it do some processing.
+    input_stream = ssc.queueStream(rdd_queue)
+
+    model = StreamingTest()
+    test_result = model.registerStream(input_stream)
+
+    test_result.pprint()
+
+    ssc.start()
+    time.sleep(12)
```
Comment (feynmanliang) [inline examples/src/main/python/mllib/streaming_test_example.py]: Is this necessary? Doesn't seem to be required for `streaming_kmeans` example

File: mllib/src/main/scala/org/apache/spark/mllib/api/python/PythonMLLibAPI.scala [hunk 7]
```diff
@@ -44,7 +44,7 @@ import org.apache.spark.mllib.regression._
 import org.apache.spark.mllib.stat.{KernelDensity, MultivariateStatisticalSummary, Statistics}
 import org.apache.spark.mllib.stat.correlation.CorrelationNames
 import org.apache.spark.mllib.stat.distribution.MultivariateGaussian
-import org.apache.spark.mllib.stat.test.{ChiSqTestResult, KolmogorovSmirnovTestResult}
+import org.apache.spark.mllib.stat.test._
```
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/api/python/PythonMLLibAPI.scala]: Why are we using a wildcard import here?

File: mllib/src/main/scala/org/apache/spark/mllib/api/python/PythonMLLibAPI.scala [hunk 8]
```diff
@@ -1358,6 +1363,41 @@ private[spark] abstract class SerDeBase {
...
   }
+
+  /**
+   * Convert a DStream of Java objects to a DStream of serialized Python objects, that is usable by
+   * PySpark.
+   */
+  def javaToPython(jDStream: JavaDStream[Any]): JavaDStream[Array[Byte]] = {
+    val dStream = jDStream.dstream.mapPartitions { iter =>
+      initialize()  // let it called in executor
+      new SerDeUtil.AutoBatchedPickler(iter)
```
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/api/python/PythonMLLibAPI.scala]: This should be an `Array[Byte]`, so you don't need to assign to `dStream` and then call the constructor on L1376

File: mllib/src/main/scala/org/apache/spark/mllib/api/python/PythonMLLibAPI.scala [hunk 9]
```diff
@@ -1358,6 +1363,41 @@ private[spark] abstract class SerDeBase {
...
+          obj match {
+            case list: JArrayList[_] => list.asScala
+            case arr: Array[_] => arr
+          }
+        } else {
+        Seq(obj)
+        }
+      }
+    }
+    new JavaDStream[Any](dStream)
```
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/api/python/PythonMLLibAPI.scala]: No need for this constructor, just return `dStream` (in fact, don't even need the `dStream` variable)

File: python/pyspark/mllib/stat/test.py [hunk 10]
```diff
@@ -15,10 +15,15 @@
...
 
+from collections import namedtuple
+
+from pyspark import SparkContext, since
 from pyspark.mllib.common import inherit_doc, JavaModelWrapper
+from pyspark.streaming.dstream import DStream
 
 
-__all__ = ["ChiSqTestResult", "KolmogorovSmirnovTestResult"]
+__all__ = ["ChiSqTestResult", "KolmogorovSmirnovTestResult", "BinarySample", "StreamingTest",
```
Comment (feynmanliang) [inline python/pyspark/mllib/stat/test.py]: nit: alphabetize

File: python/pyspark/mllib/stat/test.py [hunk 11]
```diff
@@ -80,3 +85,118 @@ class KolmogorovSmirnovTestResult(TestResult):
...
+        Update test method
+        :param testMethod:
+          Currently supported tests: `welch`, `student`.
+        """
+        assert(testMethod in ("welch", "student"),
+               "Currently supported tests: \"welch\", \"student\"")
+        self._testMethod = testMethod
+
+    @since('2.0.0')
+    def registerStream(self, data):
```
Comment (feynmanliang) [inline python/pyspark/mllib/stat/test.py]: nit: data -> dstream

File: python/pyspark/mllib/stat/test.py [hunk 12]
```diff
@@ -80,3 +85,118 @@ class KolmogorovSmirnovTestResult(TestResult):
...
+        """
+        Register a data stream to get its test result.
+
+        :param data:
+          The input data stream, each element is a BinarySample instance.
+        """
+        self._validate(data)
+        sc = SparkContext._active_spark_context
+
+        streamingTest = sc._jvm.org.apache.spark.mllib.stat.test.StreamingTest()
```
Comment (feynmanliang) [inline python/pyspark/mllib/stat/test.py]: Why did we not define a `StreamingTestModel` and assign an instance to `_model` (like in streaming K means)?

File: python/pyspark/mllib/tests.py [hunk 13]
```diff
@@ -1688,6 +1689,44 @@ def test_binary_term_freqs(self):
                                    ": expected " + str(expected[i]) + ", got " + str(output[i]))
 
 
+class StreamingTestTest(PySparkStreamingTestCase):
```
Comment (feynmanliang) [inline python/pyspark/mllib/tests.py]: Why not `MLLibStreamingTestCase`?

File: python/pyspark/mllib/tests.py [hunk 14]
```diff
@@ -1688,6 +1689,44 @@ def test_binary_term_freqs(self):
...
+    def test_streaming_test_result_and_model(self):
+        """
+        Assert the StreamingTest return valid result, and the set method of it.
+        """
+
+        checkpoint_path = tempfile.mkdtemp()
+        self.ssc.checkpoint(checkpoint_path)
```
Comment (feynmanliang) [inline python/pyspark/mllib/tests.py]: Is this necessary?

File: python/pyspark/mllib/tests.py [hunk 15]
```diff
@@ -1688,6 +1689,44 @@ def test_binary_term_freqs(self):
...
+        rdd_queue = []
+        for i in range(5):
+            rdd_queue += [self.ssc.sparkContext.parallelize(
+                [BinarySample(True, j) for j in range(1, 1001)], 10)]
+
+        # Create the QueueInputDStream and use it do some processing.
+        input_stream = self.ssc.queueStream(rdd_queue)
+
+        model = StreamingTest()
+        model.setPeacePeriod(1)
```
Comment (feynmanliang) [inline python/pyspark/mllib/tests.py]: Can we break this into another test just for model params like https://github.com/apache/spark/blob/39e2bad6a866d27c3ca594d15e574a1da3ee84cc/python/pyspark/mllib/tests.py#L1165?

File: mllib/src/main/scala/org/apache/spark/mllib/api/python/PythonMLLibAPI.scala [hunk 16]
```diff
@@ -1576,17 +1616,49 @@ private[spark] object SerDe extends SerDeBase with Serializable {
...
+
+    def construct(args: Array[AnyRef]): AnyRef = {
+      if (args.length != 2) {
+        throw new PickleException("should be 2")
+      }
+      BinarySample(args(0).asInstanceOf[Boolean], args(1).asInstanceOf[Double])
+    }
+  }
+
+  private[python] class StreamingTestResultPickler extends BasePickler[StreamingTestResult] {
```
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/api/python/PythonMLLibAPI.scala]: Do we need to test these in `PythonMLLibAPISuite`?

File: python/pyspark/mllib/stat/test.py [hunk 17]
```diff
@@ -80,3 +85,118 @@ class KolmogorovSmirnovTestResult(TestResult):
...
+        streamingTest.setPeacePeriod(self._peacePeriod)
+        streamingTest.setWindowSize(self._windowSize)
+        streamingTest.setTestMethod(self._testMethod)
+
+        javaDStream = sc._jvm.SerDe.pythonToJava(data._jdstream, True)
+        testResult = streamingTest.registerStream(javaDStream)
```
Comment (feynmanliang) [inline python/pyspark/mllib/stat/test.py]: Why do we need `pythonToJava` and `javaToPython`; its not used for streaming K means https://github.com/apache/spark/blob/39e2bad6a866d27c3ca594d15e574a1da3ee84cc/python/pyspark/mllib/clustering.py#L773

File: mllib/src/main/scala/org/apache/spark/mllib/stat/test/TestResult.scala [hunk 18]
```diff
@@ -114,7 +114,7 @@ class KolmogorovSmirnovTestResult private[stat] (
  * Object containing the test results for streaming testing.
  */
 @Since("1.6.0")
-private[stat] class StreamingTestResult @Since("1.6.0") (
+class StreamingTestResult @Since("1.6.0") (
```
Comment (feynmanliang) [inline mllib/src/main/scala/org/apache/spark/mllib/stat/test/TestResult.scala]: Does this need to be public? Java API doesn't seem to require it

**Review by feynmanliang — CHANGES_REQUESTED** [review]
Did a first pass.

It's been awhile since I've looked at PySpark so I may be a bit rusty on some things.

**Comment by HyukjinKwon** [comment]
Hi @yinxusen, are you able to proceed this further? If not, it seems it might be better closed for now.

**Pull request closed without merging**